AdminUi/src/AdminUi/ClientApp/src/app/components/client/client-edit/client-edit.component.ts

```diff
@@ -144,16 +163,30 @@ export class ClientEditComponent {
     updateClient(): void {
         this.loading = true;
 
-        let newTierId = "";
+        let defaultTier = "";
         this.tierList.forEach((tier) => {
-            if (tier.id == this.client.tierId) {
-                newTierId = tier.id;
+            if (tier.id == this.client.defaultTier) {
+                defaultTier = tier.id;
                 return;
             }
         });
```

Comment:
So I had this scenario in my head, where if we had an invalid DefaultTier associated with a Client (if for example the Tier no longer exists), then the select default tier would appear to have no tier selected as there is no match for this particular Tier. The user would then press the save button, hoping it would use the Basic Tier, but would be met with an error.

This logic behind this loop is the following:
If there isn't a Tier that matches the DefaultTier in the Tier list, then it discards the DefaultTier and uses the empty string defaultTier.
Seeing the defaultTier empty, the loop below would assume no Tier was selected and would use the Basic Tier.
If it does find the DefaultTier in the list, it uses that instead.

I tested this by deleting a Tier and attempting to change the DefaultTier of a Client.
Normally this behaviour shouldn't happen, but it seems the EntityTypeConfiguration does not assume the DefaultTier as a ForeignKey. Should I add the DefaultTier as a ForeignKey to the CustomOpenIddictCoreApplication?

And what's the expected behaviour when deleting a Tier? Should the deletion be denied or should the deletion go through and the Basic Tier become the new DefaultTier for those Applications?

Reply:
1. If I understand you correctly, the problem you describe can only occur when the following steps are executed in order: 
    1. User1 selects a new default Tier for a client
    2. User2 (or User1 in another tab) deletes the selected Tier
    3. User1 clicks on save
    This will happen once in a lifetime. So we can definitely ignore this in the frontend. Of course, the backend should (and I think already does) handle the case of an inexistent Tier.
2. Yes, you should establish a foreign key relationship on Client.DefaultTier and Tier.Id.
3. It should not be possible to delete a Tier that is used as a default Tier somewhere.

Reply:
Understood.
I added a ForeignKey restraint on the last commit, with a DeleteBehavior.Restrict , so it doesn't allow queries that attempt to remove a Tier with a relation to a Client.
In addition to this, I added a number of clients verification to the Delete Client handler, with a corresponding error.